proto/substrait/algebra.proto

```diff
@@ -427,6 +433,16 @@ message Expression {
       // A homogeneously typed list of literals
       repeated Literal values = 1;
     }
+
+    message UserDefined {
+      // points to a type_anchor defined in this plan
+      uint32 type_reference = 1;
+
+      oneof data {
+        bytes raw = 2;
+        google.protobuf.Any ext = 3;
```

Comment:
Doesn't protobuf's serialization and deserialization special-case Any to automatically serialize/deserialize the contents of that bytes field via the protobuf message type also specified therein? I don't actually know, but I assumed that's how it works.

Reply:
I don't think so. It seems you [need the internet](https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/any.proto#L133-L145) to use them 🤷🏻

Reply:
My experience is it's entirely up to the language implementation how they are handled. In java, you can register the extensions so the apis mostly handle things for you.

I haven't seen any implementation actually go to the internet to get any `Any` definitions. I haven't specifically looked for it, either.

Reply:
Yeah, sorry I didn't mean you literally have to go to the public internet, but that the comments suggests it's common for there to be an `Any` server 🤣

Reply:
Oh dear, I thought by `url` they meant something like `substrait.Plan`. That's *impressively* heavyweight.

Still... use of Any implies to me that the enclosed bytes at least correspond to the binary serialization of *some* protobuf message. I also remember having a discussion with someone who wasn't aware you could even access the contents of an Any message with the protobuf APIs directly, and *had* to deserialize it somehow. My intention for the separation was that `raw` can be used for any raw serialization, like the binary field in a decimal literal. But I don't feel super strongly about this.

Reply:
My thinking is that a raw `bytes` field has virtually no chance of being consumable by anything except the thing that produced it, whereas at least with `Any` there's a *shot*.

Reply:
Well... in order for user-defined types to be meaningful, the producer and consumer already need to agree on using the same extension. IMO the interpretation of `raw` is therefore already fixed by the extension URI and type (via `type_reference`), and protobuf encapsulation would just be a bonus if it makes sense to use that. I mean, if someone adds an i128 type, it wouldn't be very sensible to make a proto message containing a single bytes field, serialize that, and then stick that serialization in a bytes field.

Reply:
|it wouldn't be very sensible to make a proto message containing a single bytes field, serialize that, and then stick that serialization in a bytes field.

That feels pretty subjective. Being explicitly descriptive seems generally positive. Since that's exactly how we would represent that data in the canonical case, I don't see why we wouldn't ask extenders to use the same pattern.

I'd expect most plan have a handful of these values (with 10s of thousands of these values in a plan at most).

Reply:
FYI, I don't actually have a super strong opinion of whether we should allow the raw bytes field. My gut says it isn't in alignment with the "extensibility with discipline" goal of the project but I'll defer to @cpcloud and @jvanstraten.

Reply:
I'd argue that "this is written using the raw binary representation of this type" isn't any less explicit than "this is written using the raw binary representation of this type, encapsulated in a protobuf message, encapsulated in an Any," the latter is just more verbose. It would save a lot of boilerplate and error checking code for consumers and producers that don't need the complex serializations supported by protobuf; a consumer could just reject plans with the `ext` field, throw the `raw` field into pre-existing deserialization code, and call it a day. Conversely, if only `ext` would exist, they'd probably just not populate the type URL field to accomplish the same thing, which is IMO *less* explicit, and also incorrect usage of Any.

I'm not really worried about the runtime serialization overhead or size overhead, because as you say, there really shouldn't be a lot of these in a plan (or if there are, this is probably the least of your concerns).

Would it help if I added some comments? Something like

```
// the value of the literal, serialized in a format dictated by the extension
bytes raw = 2;
// the value of the literal, using protobuf serialization
google.protobuf.Any ext = 3; 
```

Reply:
If we're going to support both `raw` and `ext`, then it should probably be a `oneof`. Without that, someone can fill in both and then the desired field to consume is ambiguous.

Reply:
I agree completely, that's why it's already in a oneof :)

Reply:
Ha definitely missed that. Awesome.

Reply:
Thinking more about this I think the effort to use `raw` and `Any` is roughly the same, with `Any` providing a tiny bit more decoupling by using a type URL.

If the two representations offer equivalent functionality I'd prefer we have just one reprsentation so that consumers don't need to handle both cases.

Reply:
I still don't really agree, but I'll concede because only supporting `Any` for now would make adding `raw` later a new feature, whereas the other way around would be breaking. Let's just see how it plays out in practice.